app/services/preserver/importer.rb

```diff
@@ -0,0 +1,121 @@
+      file_metadata = resource.metadata_node
+      metadata_file_identifier = if file_metadata.nil?
+                                   nil
+                                 else
+                                   file_metadata.file_identifiers.first
+                                 end
+      binary_nodes = resource.binary_nodes
+      binary_file_identifiers = binary_nodes.map(&:file_identifiers)
+      binary_file_identifiers.flatten!
+
+      resource_storage_adapter = storage_adapter || default_storage_adapter
+      instance = new(
+        metadata_file_identifier: metadata_file_identifier,
+        binary_file_identifiers: binary_file_identifiers,
+        change_set_persister: change_set_persister,
+        storage_adapter: resource_storage_adapter
+      )
+      instance.import!
+    end
+
+    def initialize(metadata_file_identifier:, binary_file_identifiers:, change_set_persister:, storage_adapter: nil)
+      @metadata_file_identifier = metadata_file_identifier
+      @binary_file_identifiers = binary_file_identifiers
+      @storage_adapter = storage_adapter || default_storage_adapter
+      @change_set_persister = change_set_persister
+    end
+
+    def import!
+      fs_attributes = import_metadata(metadata_file_identifier)
+
+      files = import_binary_nodes(binary_file_identifiers)
+      fs_attributes[:files] = files unless files.empty?
+      fs = FileSet.new
+      fs_change_set = FileSetChangeSet.new(fs)
+      fs_change_set.validate(fs_attributes)
+      persisted = nil
+      change_set_persister.buffer_into_index do |buffered_change_set_persister|
+        persisted = buffered_change_set_persister.save(change_set: fs_change_set)
+      end
+      persisted
+    end
+
+    private
+
+      def build_id(json)
+        Valkyrie::ID.new(json["id"])
+      end
+
+      def build_optimistic_lock_token(json)
+        return unless json.key?("adapter_id")
+
+        adapter_id = build_id(json["adapter_id"])
+        token = json["token"]
+
+        Valkyrie::Persistence::OptimisticLockToken.new(
+          adapter_id: adapter_id,
+          token: token
+        )
+      end
+
+      def build_file_set_attributes(json)
+        return unless json.key?("id")
+
+        id = build_id(json["id"])
+        json["id"] = id
+
+        optimistic_lock_tokens = json["optimistic_lock_token"].map do |lock_json|
+          build_optimistic_lock_token(lock_json)
+        end
+        json["optimistic_lock_token"] = optimistic_lock_tokens
+
+        fs_attributes = json.symbolize_keys
+        fs_attributes.delete(:file_metadata)
+        fs_attributes
+      end
+
+      def import_metadata(file_identifier)
+        return {} if file_identifier.nil?
+        file = storage_adapter.find_by(id: file_identifier)
+
+        file_contents = file.read
+        metadata_json = JSON.parse(file_contents)
+        build_file_set_attributes(metadata_json)
```

Comment:
I think you can use the conversion routines which are a part of Valkyrie. I can't really dig into this right now, but instead of manually parsing the metadata maybe you can look at using https://github.com/samvera/valkyrie/blob/2-0-rc/lib/valkyrie/persistence/shared/json_value_mapper.rb when 2.0 is released? Or even just pass it through the valkyrie-sequel ResourceFactory?